packages/vm/src/middleware/deterministic.rs

```diff
@@ -1,210 +1,201 @@
+            | Operator::I32DivS
+            | Operator::I32DivU
+            | Operator::I32RemS
+            | Operator::I32RemU
+            | Operator::I32And
+            | Operator::I32Or
+            | Operator::I32Xor
+            | Operator::I32Shl
+            | Operator::I32ShrS
+            | Operator::I32ShrU
+            | Operator::I32Rotl
+            | Operator::I32Rotr
+            | Operator::I64Clz
+            | Operator::I64Ctz
+            | Operator::I64Popcnt
+            | Operator::I64Add
+            | Operator::I64Sub
+            | Operator::I64Mul
+            | Operator::I64DivS
+            | Operator::I64DivU
+            | Operator::I64RemS
+            | Operator::I64RemU
+            | Operator::I64And
+            | Operator::I64Or
+            | Operator::I64Xor
+            | Operator::I64Shl
+            | Operator::I64ShrS
+            | Operator::I64ShrU
+            | Operator::I64Rotl
+            | Operator::I64Rotr
+            | Operator::I32WrapI64
+            | Operator::I32Extend8S
+            | Operator::I32Extend16S
+            | Operator::I64Extend8S
+            | Operator::I64Extend16S
+            | Operator::I64ExtendI32S
+            | Operator::I64ExtendI32U => {
+                state.push_operator(operator);
+                Ok(())
+            }
+            _ => {
+                // Dirty hack: create an BinaryReaderError
+                // See https://github.com/wasmerio/wasmer/issues/1950
+                let mut reader = BinaryReader::new(b"\0\0\0\0");
+                let error = reader.read_bytes(10).unwrap_err();
+                Err(error)
+            }
+        }
     }
 }
 
-/// Middleware is only supported in singlepass backend, see
-/// https://github.com/CosmWasm/cosmwasm/issues/311
 #[cfg(test)]
 mod tests {
-    // No 'use super::*;' here. This is strange and means we are not testing the functions in this module directly.
-    use crate::backends::compile;
-    use crate::errors::VmError;
-    use wasmer_runtime_core::{imports, typed_func::Func};
+    use super::*;
+    use std::sync::Arc;
+    use wasmer::{CompilerConfig, Cranelift, Module, Store, JIT};
 
     #[test]
     fn valid_wasm_instance_sanity() {
-        let input = r#"
+        let wasm = wat::parse_str(
+            r#"
             (module
                 (func (export "sum") (param i32 i32) (result i32)
                     get_local 0
                     get_local 1
                     i32.add
                 ))
-            "#;
-        let wasm = wat::parse_str(input).unwrap();
-        let module = compile(&wasm).unwrap();
-        let instance = module.instantiate(&imports! {}).unwrap();
+            "#,
+        )
+        .unwrap();
 
-        let func: Func<(i32, i32), i32> = instance.exports.get("sum").unwrap();
-        let res = func.call(10, 20);
-        assert!(res.is_ok());
-        assert_eq!(30, res.unwrap());
+        let deterministic = Arc::new(Deterministic::new());
```

Comment:
Nice to see these steps. They were previously burried, but it actually reads quite well now to construct such middleware stack